### PR TITLE
Add map folding tutorial

### PR DESCRIPTION
**Pull request by TristanCacqueray**
This change adds a tutorial to implement map and fold to generate
list of text.

### CODE REVIEW

**Comment by TristanCacqueray** [comment]
This type of doc should probably go elsewhere, though, I don't know where. I'm posting it here in the meantime. Also, I am new to this language and the code may be quite inefficient, but I figure this might be useful content for a new comer.

**Comment by Gabriel439** [comment]
In https://discourse.dhall-lang.org/t/contributing-to-the-wiki-documentation/107/3?u=gabriel439 I came to the conclusion that we can host documentation in this repository and eventually be able to expose it via the dhall-lang.org website underneath a docs.dhall-lang.org subdomain.  This subdomain would eventually supersede the GitHub wiki

So what that means is that I'm in favor of adding documentation here for now and then I'll work on having the NixOps deploy take care of automatically rendering the documentation form markdown to HTML with some documentation building tool.

File: tutorials/string-matrix.md [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+that reduces a list of objects to another type. Its signature is:
+
+* input type (A),
+* the list to fold (a List of A),
+* output type (B),
+* a folding procedure that goes from (A, B) to B,
+* a starting value of type B, and
+* return a B.
+
+Therefore, we can implement a `fold-list-text` procedure like so:
```
Comment (Gabriel439) [inline tutorials/string-matrix.md]: It might be worth mentioning that this can also be defined as:

```
fold-list-text = https://prelude.dhall-lang.org/List/concat Text
```

File: tutorials/string-matrix.md [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+in  generate-test-suite [ "suffixA", "suffixB" ] [ "variantA", "variantB" ]
+```
+
+This is equivalent to the previous evaluation, but the fold-list-text procedure is now hidden in
+the generate-test-suite thanks to a nested `let`.
+
+We could use the new procedure to generate even more variations:
+
+```dhall
+generate-test-suite
```
Comment (Gabriel439) [inline tutorials/string-matrix.md]: It might be worth changing the argument order to `generate-test-suite` so that the prefixes come first and suffixes come fsecond

File: tutorials/string-matrix.md [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+[Map](https://en.wikipedia.org/wiki/Map_%28higher-order_function%29) is a procedure
+that applies a method to each element of a list, for example:
+
+```dhall
+let List/map =
+      https://raw.githubusercontent.com/dhall-lang/dhall-lang/v11.1.0/Prelude/List/map sha256:dd845ffb4568d40327f2a817eb42d1c6138b929ca758d50bc33112ef3c885680
+
+let generate-test-name =
+        λ(suffix : Text)
```
[the remark this answers is not in the document]
Comment (TristanCacqueray) [reply tutorials/string-matrix.md]: I also got that criticism, e.g. "how do you type λ?"... Finally figured out how to make the emacs dhall-mode format the code using ascii.

File: tutorials/string-matrix.md (hunk 1 above)
Comment (TristanCacqueray) [reply tutorials/string-matrix.md]: Ah, I didn't knew about that function :) I guess it's still relevant to go through simple implementation of prelude functions.

File: tutorials/string-matrix.md [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+``` yaml
+- test-variantA-suffixA
+- test-variantA-suffixB
+- test-variantB-suffixA
+- test-variantB-suffixB
+```
+
+# Simple map
+
+[Map](https://en.wikipedia.org/wiki/Map_%28higher-order_function%29) is a procedure
```
Comment (amarrella) [inline tutorials/string-matrix.md]: nitpick: here and below could we say `function` instead of `procedure` and `method`?

File: tutorials/string-matrix.md [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+- variantB-suffixA
+```
+
+In dhall, generic methods such as `map` require abstract types to be explicit. The map signature is:
+
+* input type (A),
+* output type (B),
+* a procedure that goes from A to B,
+* a list of A, and
+* return a list of B.
```
[the remark this answers is not in the document]
Comment (TristanCacqueray) [reply tutorials/string-matrix.md]: I find it quite difficult to explain `List/map` usage for someone not familiar with the currying and partial function application. I wonder if this need a tutorial on its own, e.g. `define a function that takes two arguments` and perhaps a discussion about the benefit of currying technique.

File: tutorials/string-matrix.md [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+Thus, the generate-test-name pre applies the `map` method from `Text` to `Text` using a known suffix.
+Then it is given a list of prefixes to generate new strings.
+We could have inverted the suffix and prefix, but this will useful later.
```
[the remark this answers is not in the document]
Comment (TristanCacqueray) [reply tutorials/string-matrix.md]: I changed the code so that prefixes come before suffixes.

File: tutorials/string-matrix.md (hunk 4 above)
Comment (TristanCacqueray) [reply tutorials/string-matrix.md]: Good point

**Comment by TristanCacqueray** [comment]
Thank you for the prompt feedback, I really appreciate this :)
I left a couple of things to do I'd like to resolve once the new docs website is available.

File: docs/tutorials/string-matrix.md [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+      https://prelude.dhall-lang.org/v11.1.0/List/map sha256:dd845ffb4568d40327f2a817eb42d1c6138b929ca758d50bc33112ef3c885680
+
+let generate-test-name =
+          \(prefix : Text)
+      ->  List/map Text Text (\(suffix : Text) -> "${prefix}-${suffix}")
+
+in  generate-test-name "variantA" [ "suffixA", "suffixB" ]
+```
+```console
+$ dhall-to-yaml --file ./string-matrix.dhall
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]: I would also include the output of the `dhall --file` command, too, for users less familiar with YAML:

```bash
$ dhall --file ./string-matrix.dhall
[ "variantA-suffixA", "variantA-suffixB" ]
```

File: docs/tutorials/string-matrix.md [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+- variantA-suffixA
+- variantA-suffixB
+```
+
+In dhall, generic functions such as `map` require abstract types to be explicit. The `map` signature is:
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]:
```suggestion
In dhall, generic functions such as `map` require types to be explicitly specified as function arguments. The `map` signature is:
```

File: docs/tutorials/string-matrix.md [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+
+* input type (A),
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]: I would include the full type signature for `map` here before explaining the components of the type

File: docs/tutorials/string-matrix.md [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+List/map Text Text (\(prefix : Text) -> "${prefix}-${suffix}")
+    : List Text -> List Text
+```
+
+Thus, `generate-test-name` pre-applies the `map` function from `Text` to `Text` using a known prefix.
+Then it is given a list of suffixes to generate new strings.
+
+## Double map
+
+We can call two `map`s and generate our initial goal by adding such a function to our `string-matrix.dhall` file:
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]:
```suggestion
We can call two `map`s and generate our initial goal by adding the following function to our `./string-matrix.dhall` file:
```

File: docs/tutorials/string-matrix.md [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+      ->  List/map
+            Text
+            (List Text)
+            (\(prefix : Text) -> generate-test-name prefix suffixes)
+            prefixes
+
+in  generate-test-suite [ "variantA", "variantB" ] [ "suffixA", "suffixB" ]
+```
+```console
+$ dhall-to-yaml --file ./string-matrix.dhall
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]: Same suggestion here: also include the output of the `dhall` command for reference

File: docs/tutorials/string-matrix.md [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+- - variantB-suffixA
+  - variantB-suffixB
+```
+
+However, by doing so we get a value of type `List (List Text)` (i.e. a `List` of `List`s of `Text`).
+So we can get a flat list by using the fold function.
+
+## Folding a list of list of texts to a flat list of text
+
+[Fold](https://en.wikipedia.org/wiki/Fold_%28higher-order_function%29) is a function
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]:
```suggestion
A ["fold"](https://en.wikipedia.org/wiki/Fold_%28higher-order_function%29) is a function
```

File: docs/tutorials/string-matrix.md [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+that reduces a list of objects to another type. Its signature is:
+
+* input type (A),
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]: Same suggestion here: include the full type signature for `List/fold` first before explaining the components of the type

File: docs/tutorials/string-matrix.md [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+* the list to fold (a List of A),
+* output type (B),
+* a folding function that goes from (A, B) to B,
+* a starting value of type B, and
+* return a B.
+
+Therefore, we can implement a `fold-list-text` function like so:
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]: Perhaps call this function "`flatten-list-text`" since once you saturate the `List/fold` function the result is no longer a "fold"

File: docs/tutorials/string-matrix.md [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+The [fold](https://prelude.dhall-lang.org/List/fold) documentation actually mentions:
+
+> If you treat the list `[ x, y, z ]` as `cons x (cons y (cons z nil))`, then a
+> `fold` just replaces each `cons` and `nil` with something else.
+
+This is exactly what we need, `cons` is our `generate-test-suite` and `nil` is
+the initial suffix `[ "el6", "el7" ]`!
+
+```dhall
+let generate-test-matrix =
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]: It seems like it would be more elegant if the `generate-test-matrix` function had the following type:

```dhall
generate-test-matrix : List (List Text) -> List Text
```

... so that you could then write:

```dhall
generate-test-matrix [ [ "tox" ], ["py27", "py38" ], [ "postgresql", "sqlite" ], [ "el6", "el7" ] ]
```

Would it be possible to extend the tutorial to show how to do that?

File: docs/tutorials/string-matrix.md [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+- tox-py38-sqlite-el6
+- tox-py27-postgresql-el7
+- tox-py38-postgresql-el7
+- tox-py27-sqlite-el7
+- tox-py38-sqlite-el7
+```
+
+> **Note**: Using the `List/head` function the `generate-test-matrix` could
+> be improved to take a single list of argument.
+> A follow-up tutorial introduces that using `Optional/fold`
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]: Is this note up-to-date?  I couldn't find the `List/head` that this was referring to

File: docs/tutorials/string-matrix.md [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+
+In dhall, generic functions such as `map` require abstract types to be explicit. The `map` signature is:
+
+* input type (A),
+* output type (B),
+* a function that goes from A to B,
+* a list of A, and
+* return a list of B.
+
+Functions can be pre-applied as each argument evaluates to a new function.
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]:
```suggestion
Functions can be partially saturated as each argument evaluates to a new function.
```

File: docs/tutorials/string-matrix.md [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+This technique is called [currying](https://en.wikipedia.org/wiki/Currying).
+For example, the above `List/map` call successively evaluates as:
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]:
```suggestion
For example, the type of the `List/map` function changes with each new argument:
```

File: docs/tutorials/string-matrix.md (hunk 14 above)
Comment (TristanCacqueray) [reply docs/tutorials/string-matrix.md]: Good point.

File: docs/tutorials/string-matrix.md (hunk 15 above)
Comment (TristanCacqueray) [reply docs/tutorials/string-matrix.md]: It seems like we either have to use `List/last` (and introduce optionals), or rewrite the `generate-test-suite` function to handle empty initial prefixes. I meant to implement the first solution in another tutorial doc to keep it simple, but this can be added to this doc instead.

File: docs/tutorials/string-matrix.md (hunk 16 above)
Comment (TristanCacqueray) [reply docs/tutorials/string-matrix.md]: Yeah, the improvement still needs to be written... Perhaps from a tutorial point of view we could explain why folding with `generate-test-suite` needs a starting prefix and that `generate-test-matrix` could be improved by using `List/last` as an exercise left to the reader. But I found `Optional/fold` to be unintuitive to use and figured this should be introduced in its own tutorial.

I meant to cover a couple more topics such as using List/filter to blacklist some combination (e.g. skip py3 on el6) and rendering more complex structure such as
```yaml
- job:
    name: tox-py38-sqlite
    vars:
      version: py36
      backend: sqlite
```
I think that would make a nice 4 parts tutorial.

**Comment by TristanCacqueray** [comment]
Thank you for the suggestions and editing my english :)

File: docs/tutorials/string-matrix.md [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+
+In dhall, generic functions such as `map` require types to be explicitly specified as function arguments. The `map` signature is:
+
+```dhall
+let map : forall (a : Type) -> forall (b : Type) -> (a -> b) -> List a -> List b
+```
+
+In other words, the `map` arguments are:
+
+* input type (A),
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]:
```suggestion
* input type (`a`),
```

For consistency with the type signature, I suggest using lowercase when referencing the types (and format the symbols as code)

File: docs/tutorials/string-matrix.md [hunk 20]
```diff
@@ -0,0 +1,262 @@
...
+
+Thus, `generate-test-name` pre-applies the `map` function from `Text` to `Text` using a known prefix.
+Then it is given a list of suffixes to generate new strings.
+
+## Double map
+
+We can call two `map`s and generate our initial goal by adding the following function to our `./string-matrix.dhall` file:
+
+```dhall
+let generate-test-suite =
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]: I suggest including the entire program at each step so that the user can follow along at each stage by copying and pasting the entire example

The other reason I suggest this is because the functions have similar names (e.g. `generate-*`) so including both of them side-by-side visually indicates that they are now working with two distinct functions instead of amending the old `generate-test-name` function)

File: docs/tutorials/string-matrix.md [hunk 21]
```diff
@@ -0,0 +1,262 @@
...
+- test-variantB-suffixA
+- test-variantB-suffixB
+```
+
+We'll use the `dhall` and `dhall-to-yaml` command line,
+make sure to follow the [install instructions][installation] from the getting started tutorial.
+
+[installation]: <tutorials/Getting-started_Generate-JSON-or-YAML:installation>
+
+## Simple map
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]:
```suggestion
## Map
```

I try to avoid the use of the word "simple" in tutorial material, since not all users might consider the concepts introduced within a tutorial to be simple

File: docs/tutorials/string-matrix.md [hunk 22]
```diff
@@ -0,0 +1,262 @@
...
+    : (Text -> Text) -> List Text -> List Text
+
+List/map Text Text (\(prefix : Text) -> "${prefix}-${suffix}")
+    : List Text -> List Text
+```
+
+Thus, `generate-test-name` pre-applies the `map` function from `Text` to `Text` using a known prefix.
+Then it is given a list of suffixes to generate new strings.
+
+## Double map
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]:
```suggestion
## Nested map
```

This is a minor suggestion to indicate that the `map`s are not being applied in succession (e.g. `map f (map g xs)`) but are instead being applied in a manner analogous to a nested `for` loop

File: docs/tutorials/string-matrix.md [hunk 23]
```diff
@@ -0,0 +1,262 @@
...
+```
+
+We'll use the `dhall` and `dhall-to-yaml` command line,
+make sure to follow the [install instructions][installation] from the getting started tutorial.
+
+[installation]: <tutorials/Getting-started_Generate-JSON-or-YAML:installation>
+
+## Simple map
+
+A ["map"](https://en.wikipedia.org/wiki/Map_%28higher-order_function%29) is a function
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]: I think this should be preceded by a paragraph motivating why we want to apply a function to each element of a list.  The reason to do so is that you want to put the reader in a frame of mind where "If I have a goal X then the solution is Y" so that the next time they have the same goal X they immediately remember to use solution Y.

Currently the prose jumps straight to the solution Y (e.g. "use a combinator that applies a function to each element of the list") without explaining why we need such a function.  Here's an example of how I might motivate the use of `map`:

> We would like to begin by prepending the same prefix to a list of suffixes, so that our final code looks like this:
>
> ```dhall
> generate-test-name "variantA" [ "suffixA", "suffixB" ]
> 
> = [ "variantA-suffixA", "variantA-suffixB" ]
> ```
>
> ... but in order to do so we need a way to apply the same transformation (e.g. append a suffix) to each element of a list.
>
> The ["map"](…) higher-order function provides precisely this capability: we can use `map` to uniformly transform each element of a list, like this:
>
> …

I have a similar comment for the following two sections: first explain to the user what the local goal is for that section before introducing the utility or design pattern that realizes that goal.

File: docs/tutorials/string-matrix.md [hunk 24]
```diff
@@ -0,0 +1,262 @@
...
+* output type (B),
+* a folding function that goes from (A, B) to B,
+* a starting value of type B, and
+* it returns a B.
+
+Therefore, we can implement a `flatten-list-text` function like so:
+
+```dhall
+-- concat-example.dhall
+let {- Convert [[a, b], [c, d]] to [a, b, c, d] -}
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]: You can actually turn this comment into an `assert` test! 🙂 

```dhall
let flatten-list-text =
        λ(to-fold : List (List Text))
      →   List/fold
            (List Text)
            to-fold
            (List Text)
            (λ(a : List Text) → λ(b : List Text) → a # b)
            ([] : List Text)
        : List Text

let example =
        λ(a : Text)
      → λ(b : Text)
      → λ(c : Text)
      → λ(d : Text)
      → assert : flatten-list-text [ [ a, b ], [ c, d ] ] ≡ [ a, b, c, d ]

in  flatten-list-text [ [ "A", "B" ], [ "C", "D" ] ]
```

File: docs/tutorials/string-matrix.md [hunk 25]
```diff
@@ -0,0 +1,262 @@
...
+* the list to fold (a List of A),
+* output type (B),
+* a folding function that goes from (A, B) to B,
+* a starting value of type B, and
+* it returns a B.
+
+Therefore, we can implement a `flatten-list-text` function like so:
+
+```dhall
+-- concat-example.dhall
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]:
```suggestion
-- ./concat-example.dhall
```

File: docs/tutorials/string-matrix.md [hunk 26]
```diff
@@ -0,0 +1,262 @@
...
+              (\(a : List Text) -> \(b : List Text) -> a # b)
+              ([] : List Text)
+          : List Text
+
+in  flatten-list-text [ [ "A", "B" ], [ "C", "D" ] ]
+```
+
+```console
+$ dhall-to-yaml --file ./concat-example.dhall
+[A, B, C, D]
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]: Is this output correct?  I would have expected:

```yaml
- A
- B
- C
- D
```

File: docs/tutorials/string-matrix.md [hunk 27]
```diff
@@ -0,0 +1,262 @@
...
+```yaml
+- variantA-suffixA
+- variantA-suffixB
+- variantB-suffixA
+- variantB-suffixB
+```
+
+## Refactoring the generate-test-suite function
+
+The flatten-list-text function doesn't have to be visible and the generate test-suite
```
Comment (Gabriel439) [inline docs/tutorials/string-matrix.md]:
```suggestion
The flatten-list-text function doesn't have to be a top-level function and the generate test-suite
```

File: docs/tutorials/string-matrix.md (hunk 20 above)
Comment (TristanCacqueray) [reply docs/tutorials/string-matrix.md]: That's a good idea.

File: docs/tutorials/string-matrix.md (hunk 21 above)
Comment (TristanCacqueray) [reply docs/tutorials/string-matrix.md]: Make sense

File: docs/tutorials/string-matrix.md (hunk 23 above)
Comment (TristanCacqueray) [reply docs/tutorials/string-matrix.md]: Thank you for the suggestion. By the way, I'm really digging the tutorial/how-to/discussions/references doc organization after going through dhall doc :)

File: docs/tutorials/string-matrix.md (hunk 26 above)
Comment (TristanCacqueray) [reply docs/tutorials/string-matrix.md]: That was not correct indeed.

File: docs/tutorials/string-matrix.md (hunk 24 above)
Comment (TristanCacqueray) [reply docs/tutorials/string-matrix.md]: That's really neat how such a test can be asserted. Thanks for the example.

**Comment by TristanCacqueray** [comment]
I continued to work on that topic and perhaps this tutorial could be extended with a structured output version (introduce custom types and optional usage), and a function to filter out some combinaison (introduce sum type), e.g. these draft expressions: https://gist.github.com/TristanCacqueray/155276c518754222636f9cb0dc6028bc .

In the meantime, I think this string-matrix introduction is good on its own. Thanks for all the feedback, there were useful.

**Comment by Gabriel439** [comment]
@TristanCacqueray: You're welcome! 🙂

**Comment by Profpatsch** [comment]
@TristanCacqueray Maybe we merge this first and then potentially expand it later? Would be a shame to let all this work go to waste.

**Comment by TristanCacqueray** [comment]
@Profpatsch yes, and I currently don't have time to write the follow-up tutorials. Though I'd like to propose the [docs/conf.py](https://github.com/dhall-lang/dhall-lang/pull/824/files#diff-85987f48f1258d9ee486e3191495582d) modification in another PR, and ensure this is working as expected. That could also be used to fix the remaining gfm links (e.g. `JSON tutorial` should be a link in https://docs.dhall-lang.org/discussions/Core-language-features.html ). I can do that shortly if that's ok for you.

**Comment by TristanCacqueray** [comment]
GitHub doesn't seems to support cross PR dependencies, thus we should land https://github.com/dhall-lang/dhall-lang/pull/888 first, then i'll rebase and squash the "string-matrix" tutorial.

**Comment by TristanCacqueray** [comment]
4ba5a9f squashed all the previous commit on top of master.
